firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,508 @@
+    dimension = 0 # by definition since point cloud
+
+    nodes = nodes_per_entity.reshape(dimension + 1, -1)
+
+    for i in range(dimension + 1):
+        pStart = 0 # by definition since a swarm is point cloud
+        CHKERR(DMSwarmGetLocalSize(dm.dm, &pEnd)) # by definition since a swarm is point cloud
+        if not variable:
+            ndof = nodes[i, 0]
+        for p in range(pStart, pEnd):
+            if variable:
+                if on_base_:
+                    ndof = nodes[i, 1]
+                else:
+                    layers = layer_extents[p, 1] - layer_extents[p, 0]
+                    ndof = layers*nodes[i, 0] + (layers - 1)*nodes[i, 1]
+            CHKERR(PetscSectionSetDof(section.sec, p, ndof))
+    section.setUp()
+    return section
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def closure_ordering(PETSc.DM swarm,
+                     PETSc.Section vertex_numbering,
+                     PETSc.Section cell_numbering,
+                     np.ndarray[PetscInt, ndim=1, mode="c"] entity_per_cell):
+    """Apply Fenics local numbering to a cell closure.
+
+    ..note:: This is a copy and paste job from the equivalent function
+             in dmplex.pyx with the relevant modifications made for a
+             DMSwarm.
+
+    :arg swarm: The DMSwarm object encapsulating the vertex-only mesh topology
+    :arg vertex_numbering: Section describing the universal vertex numbering
+    :arg cell_numbering: Section describing the global cell numbering
+    :arg entity_per_cell: List of the number of entity points in each dimension
+
+    Vertices    := Ordered according to global/universal
+                   vertex numbering
+    Edges/faces := Ordered according to lexicographical
+                   ordering of non-incident vertices
+    """
+    cdef:
+        PetscInt c, cStart, cEnd, v, vStart, vEnd
+        PetscInt f, fStart, fEnd, e, eStart, eEnd
+        PetscInt dim, vi, ci, fi, v_per_cell, cell
+        PetscInt offset, cell_offset, nfaces, nfacets
+        PetscInt *vertices = NULL
+        PetscInt *v_global = NULL
+        PetscInt closure
+        PetscInt *facets = NULL
+        PetscInt *faces = NULL
+        PetscInt *face_indices = NULL
+        const PetscInt *face_vertices = NULL
+        PetscInt *facet_vertices = NULL
+        np.ndarray[PetscInt, ndim=2, mode="c"] cell_closure
+
+    dim = 0 # by definition since point cloud
+    cStart = 0
+    cEnd = swarm.getLocalSize()
+    vStart = cStart
+    vEnd = cEnd
+    v_per_cell = entity_per_cell[0]
+    cell_offset = sum(entity_per_cell) - 1
+
+    CHKERR(PetscMalloc1(v_per_cell, &vertices))
+    CHKERR(PetscMalloc1(v_per_cell, &v_global))
+    cell_closure = np.empty((cEnd - cStart, sum(entity_per_cell)), dtype=IntType)
+
+    for c in range(cStart, cEnd):
+        CHKERR(PetscSectionGetOffset(cell_numbering.sec, c, &cell))
+        closure = c # by definition since just the vertex
```

Comment:
It looks like all these copy-and-paste functions could be removed if you implemented three small functions:

```python
def getChart(dm):
    if dm.getType() == dm.Type.PLEX:
        return dm.getChart()
    elif dm.getType() == dm.Type.SWARM:
        return 0, dm.getLocalSize()
    else:
       raise TypeError(...)


cdef inline get_transitive_closure(PETSc.DM dm, PetscInt p, PetscBool useCone, PetscInt *nclosure, PetscInt **closure):
    cdef PetscBool isplex, isswarm

    CHKERR(PetscObjectTypeCompare(<PETSc.PetscObject>dm.dm,PETSc.DM.Type.PLEX,&isplex))
    CHKERR(PetscObjectTypeCompare(<PETSc.PetscObject>dm.dm,PETSc.DM.Type.SWARM,&isswarm))
    if isswarm:
        if nclosure: 
            nclosure[0] = 1
        if closure[0] == NULL:
            CHKERR(PetscMalloc1(2, closure))
        closure[0][0] = p
        closure[0][1] = 0
    else:
        CHKERR(DMPlexGetTransitiveClosure(dm.dm, p, useCone, nclosure, closure))

cdef inline restore_transitive_closure(...):
    ...
    if isswarm:
        if closure != NULL:
            CHKERR(PetscFree(closure[0]))
    else:
       CHKERR(DMPlexRestoreTransitiveClosure(...)
````

Reply:
Any thoughts on this proposal?

Reply:
I don't think I understand what you are proposing. I don't see how this could replace all the copy and paste functions - there's a lot more to do than just get the transitive closure - and I haven't yet generalised mesh dms to be either plex or swarm. The latter is a pretty massive refactor that will have to happen in another branch as discussed previously.

Reply:
Maybe create `dmcommon.pyx` (or whatever) and gradually move common things there? At least some functions in `dmplex.pyx` and `dmswarm.pyx` would look identical with what @wence- suggested.

Reply:
Another approach, we can keep them separate, and _remove_ all of the generic code and just do TRT for swarm objects (with only vertices as points). I would kind of be happier with this, since then we have an implementation that isn't "fakely" generic.

Reply:
That would definitely be easier for me. I'll still generalise the mesh code to have a `_dm` property though. What does TRT stand for?

Reply:
the right thing.

Reply:
I _think_ this is now resolved. Will await further comment.

Reply:
No it's not resolved. This is a great idea and I will do it

Reply:
Done